src/client/world/chunk_mesh_generation.cpp

```diff
@@ -96,8 +96,11 @@ ChunkMeshCollection makeChunkMesh(const Chunk& chunk, const VoxelDataManager& vo
                                       voxData.bottomTextureId);
                     }
 
+                    bool renderTopIfFluid = (voxData.type == VoxelType::Fluid) &&	// Render the top face if this voxel is fluid...
+                        (voxelData.getVoxelData(chunk.getVoxel({x, y + 1, z})).type != VoxelType::Fluid);	// and this one not.
```

Comment:
> this could be moved to a seperate function

Yeah, that will also increase readability.

> Doing it this way would mean if this `makeFace...` returns false, then the second check won't be made

`makeFace` will return false if the face is covered with a block (that was the problem with the water), so the second check (that's OR'd with the first one) will be evaluated to make sure that we're not covering the top-face of a fluid.

> to avoid doing this check for every single block

The only thing we can do is to skip this check if `makeFace` returns true in the first place, since any other method will require to check if this is a water voxel, opaque block or "gapful" voxel anyway.

Fix:

```cpp
auto topVoxel = chunk.getVoxel({x, y + 1, z});    // Don't retrieve the voxel twice.

if (makeFace(voxelData, voxel, topVoxel)) || shouldRenderTop(voxData, voxelData.getVoxelData(topVoxe))) {
    ...
}

bool shouldRenderTop(VoxelData actualVoxData, VoxelData topVoxData) {
    return (actualVoxData.type == VoxelType::Fluid) ?
        (topVoxData.type != VoxelType::Fluid) : 
        false;    // Skips topVoxData check if actualVoxData is not a fluid
}
```

However, it's still an ad hoc fix for water. Ideally this would be scalable to glass or leaves, but that will require checks for every face (or [this](https://user-images.githubusercontent.com/4872133/74798893-cbc92b80-52ad-11ea-95e5-bbd9b282ff54.png) will haunt you).

Water sure is weird.